src/main/scala/org/apache/spark/shuffle/ucx/UcxWorkerWrapper.scala

```diff
@@ -0,0 +1,151 @@
+
+  override def getError: TransportError = null
+
+  override def getStats: Option[OperationStats] = Some(stats)
+}
+
+class UcxFailureOperationResult(errorMsg: String) extends OperationResult {
+  override def getStatus: OperationStatus.Value = OperationStatus.FAILURE
+
+  override def getError: TransportError = new TransportError(errorMsg)
+
+  override def getStats: Option[OperationStats] = None
+}
+
+/**
+ * Worker per thread wrapper, that maintains connection and progress logic.
+ */
+class UcxWorkerWrapper(worker: UcpWorker, transport: UcxShuffleTransport, ucxConf: UcxShuffleConf,
+                       memoryPool: MemoryPool)
+  extends Closeable with Logging {
+
+  // To keep connection map on a remote side by id rather then by worker address, which could be big.
+  // Would not need when migrate to active messages.
+  private val id: Array[Byte] = (transport.executorId + s"_${Thread.currentThread().getId}").getBytes
+  private final val connections = mutable.Map.empty[String, UcpEndpoint]
+  private var transactionId = Random.nextLong()
+  private val workerAddress = worker.getAddress
+
+  override def close(): Unit = {
+    connections.foreach{
+      case (_, endpoint) => endpoint.close()
+    }
+    connections.clear()
+    worker.close()
+  }
+
+  /**
+   * The only place for worker progress
+   */
+  private[ucx] def progress() {
+    if ((worker.progress() == 0) && ucxConf.useWakeup) {
+      worker.waitForEvents()
+      worker.progress()
+    }
+  }
+
+  private def getConnection(executorId: String): UcpEndpoint = {
+    val workerAdresses = transport.executorIdToAddress
+
+    if (!workerAdresses.contains(executorId)) {
+      // Block until there's no worker address for this BlockManagerID
+      val startTime = System.currentTimeMillis()
+      val timeout = ucxConf.conf.getTimeAsMs("spark.network.timeout", "100")
+      workerAdresses.synchronized {
+        while (workerAdresses.get(executorId) == null) {
+          workerAdresses.wait(timeout)
+          if (System.currentTimeMillis() - startTime > timeout) {
+            throw new UcxException(s"Didn't get worker address for $executorId during $timeout")
+          }
+        }
+      }
+    }
+
+    connections.getOrElseUpdate(executorId, {
+      logInfo(s"Worker from thread ${Thread.currentThread().getName} connecting to $executorId")
+      val endpointParams = new UcpEndpointParams()
+        .setUcpAddress(workerAdresses.get(executorId))
+     worker.newEndpoint(endpointParams)
+    })
+  }
+
+  def fetchBlockByBlockId(executorId: String, blockId: BlockId,
+                          resultBuffer: MemoryBlock, cb: OperationCallback): UcxRequest = {
+    val stats = new UcxStats()
+    val ep = getConnection(executorId)
+    val mem = memoryPool.get(transport.ucxShuffleConf.rpcMessageSize)
+    val buffer = UcxUtils.getByteBufferView(mem.address,
+      transport.ucxShuffleConf.rpcMessageSize.toInt)
```

Comment:
Buffer sizes are capped to 2GB? That may be an issue.

Reply:
This is for RPC message size. It's always a host buffer, that should just contain serialized BlockId